tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,219 @@
+ * the length of the next record. Instead the socket returns the minimum of
+ * bytes-requested and data-available; reading multiple consecutive records if they
+ * are of the same type. */
+ssize_t s2n_test_ktls_recvmsg_stuffer_io(struct s2n_connection *conn, struct msghdr *msg, uint8_t *record_type)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->recv_io_context);
+    POSIX_ENSURE_REF(msg);
+    POSIX_ENSURE_REF(msg->msg_iov);
+    POSIX_ENSURE_REF(record_type);
+
+    struct s2n_test_ktls_io_stuffer *io_ctx = (struct s2n_test_ktls_io_stuffer *) conn->recv_io_context;
+    POSIX_ENSURE_REF(io_ctx);
+    io_ctx->send_recv_msg_invoked_count++;
+    /* s2n only receives using msg_iovlen of 1 */
+    POSIX_ENSURE_EQ(msg->msg_iovlen, 1);
+    uint8_t *buf = msg->msg_iov->iov_base;
+    POSIX_ENSURE_REF(buf);
+
+    /* There is no data available so return blocked */
+    if (!s2n_stuffer_data_available(&io_ctx->ancillary_buffer)) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    ssize_t total_read = 0;
+    /* updated as partial or multiple records are read */
+    size_t updated_requested_len = msg->msg_iov->iov_len;
+    /* track two record_types since it is possible to read multiple records of the same type */
+    *record_type = 0;
+    uint8_t next_record_type = 0;
+    while (*record_type == next_record_type) {
+        /* read record_type and number of bytes available in the next record */
+        POSIX_GUARD(s2n_stuffer_read_uint8(&io_ctx->ancillary_buffer, record_type));
+        uint16_t n_avail = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint16(&io_ctx->ancillary_buffer, &n_avail));
+        POSIX_ENSURE_LTE(n_avail, s2n_stuffer_data_available(&io_ctx->data_buffer));
+
+        /* read minimul of requested_len and bytes_available */
+        size_t n_read = MIN(updated_requested_len, n_avail);
+        POSIX_ENSURE_GT(n_read, 0);
+
+        int ret = s2n_stuffer_read_bytes(&io_ctx->data_buffer, buf + total_read, n_read);
+        if (ret < 0) {
+            errno = EINVAL;
+            return -1;
+        }
+
+        updated_requested_len -= n_read;
+        POSIX_ENSURE_GTE(updated_requested_len, 0);
+        total_read += n_read;
+
+        /* Handle if we partially read a record */
+        ssize_t remaining_len = n_avail - n_read;
+        if (remaining_len) {
+            POSIX_GUARD_RESULT(s2n_test_ktls_update_prev_header_len(io_ctx, remaining_len));
+        }
+
+        /* if already read the requested amount then break */
+        if (updated_requested_len == 0) {
+            break;
+        }
+        /* Attempt to read multiple records (must be of the same type) */
+        if (updated_requested_len) {
+            int ret = s2n_stuffer_peek_char(&io_ctx->ancillary_buffer, (char *) &next_record_type);
+
+            bool no_more_records = ret < 0;
+            bool next_record_different_type = next_record_type != *record_type;
```

Comment:
You're sort of trying to use two conditions in order to break out of the loop.... I dunno. Why not just deal with the no_more_records case here and rely on the loop condition to break you out of the other one.